### ISSUE TITLE
Queue adapter interface is not documented or tested

### ISSUE THREAD

**Issue by liamhuber** [pyiron/pyiron_base#1085]
The queue adapter (typically accessed by [`state.queue_adapter`](https://github.com/pyiron/pyiron_base/blob/b97221d499502ba46c58a2195a4a5b236782419c/pyiron_base/state/__init__.py#L66)) got spun off to its own package, pysqa, some time ago. It's now [imported](https://github.com/pyiron/pyiron_base/blob/b97221d499502ba46c58a2195a4a5b236782419c/pyiron_base/state/queue_adapter.py#L12) and [used](https://github.com/pyiron/pyiron_base/blob/b97221d499502ba46c58a2195a4a5b236782419c/pyiron_base/state/queue_adapter.py#L58) as the standard non-None queue adapter, although in principle future devs could import a different one from somewhere else and use it.

So far this is all good. I think it's great we spun it off, and @jan-janssen has just done a flurry of activity improving test coverage on that package. What I find missing now is (a) a clear definition of the interface pyiron_base is expecting from an adapter, and (b) tests to make sure that interface is maintained. 

We recently ran into some trouble at an intermediate stage of #1082, where `state.queue_adapter.submit_job` got passed a set of `**kwargs` even though the underlying class, `pysqa.QueueAdapter.submit_jobs` doesn't accept arbitrary kwargs! This didn't raise a single red flag in pyiron_base, and I really had to go digging in the pysqa package to find the expected interface.

The queue adapter is certainly not the only place this comes up, but in general when pyiron imports some object and then uses it extensively, I think it would be nice if we documented somewhere what our expectations were for that object's interface, and, ideally, test it -- either with some proper mock up, or even just with `inspect` to verify that objects have the methods/attributes with think they do, and that the signatures we pass on calling methods are an acceptable subset of the actual method signature.

E.g. if I run a find-in-files search for `adapter.` in pyiron_base, I get 32 occurrences, all of which are for `state.queue_adapter`. I couldn't fit them all on screen, but here's a most of the list:

<img width="546" alt="Screen Shot 2023-05-04 at 10 15 09" src="https://user-images.githubusercontent.com/20283329/236285780-a31d8b22-ddd8-40c6-b9ad-c2c0f4204f4f.png">

So I would propose that somewhere in the code we have a docstring that reads something like the following:

```python
"""
Any queue adapter used should satisfy the following interface:

Takes an instantiation:
- directory (str)

Has attributes:
- remote_flag (bool)

Has methods with the following signatures:
- get_job_from_remote
  - working_directory (str)
  - delete_remote (bool) -- default False.
- ...
- submit_job:
  - queue (Optional[str])
  - job_name (Optional[str])
  - working_directory (Optional[str])
  - cores (Optional[int])
  - memory_max (Optional[int])
  - run_time_max (Optional[int])
  - dependency_list (Optional[list[str]])
  - command (Optional[str])
  - **kwargs
"""
```

I'm just providing some sparse examples here because actually filling it out would be a bit of leg work and I wanted to get feedback before putting time in (@samwaseda I know your last adventure with ChatGPT didn't end so well, but do you think this might be a task well-suited to that tool).

How/where should we clarify our interface expectations to future devs? I could see this fitting in docstrings, but also having a home in a spec doc somewhere else. How much depth is required? IMO (kw)arg name and type is already pretty helpful, even though it's minimal, but should we also say things like what the expectations are for the _behaviour_ of these methods.

This is also not something that needs to happen all at once, but as we run into classes where we're like "hey, we import this and have all sorts of implicit expectations on the imported object", it would be nice to make those expectations explicit.

Pushback? Additions?

**Comment by jan-janssen** [comment]
The question for me is where to define the interface, on the `pysqa` side we already have a rough interface definition: 
https://github.com/pyiron/pysqa#usage
This definitely needs to be extended but it at least covers the general commands used by pyiron.

**Comment by liamhuber** [comment]
Yeah, if it's fully and clearly defined there then IMO it's enough to say "any adapter must match this spec: <redirect to pysqa spec>. Although that is contingent on pysqa spec having a 1:1 correspondence to pyiron_base requirements -- we don't want to over-define/-require. 

I also think it's still wise to have tests in pyiron_base to ensure that our usage still conforms to the current state of pysqa, regardless of where we put the docs.
